File: clients/ios/SsdidDrive/SsdidDrive/Domain/Repository/ShareRepository.swift
import Foundation

/// Repository for sharing operations
protocol ShareRepository {

    // MARK: - Create Shares

    /// Share a file with a user
    /// - Parameters:
    ///   - fileId: The file ID to share
    ///   - granteeId: User ID of the recipient
    ///   - granteePublicKeys: Recipient's public keys for key wrapping
    ///   - fileEncryptedKey: The file's DEK encrypted for the owner (required for key wrapping)
    ///   - permission: Share permission level
    ///   - expiresAt: Optional expiration date
    func shareFile(
        fileId: String,
        granteeId: String,
        granteePublicKeys: KeyManager.PublicKeys,
        fileEncryptedKey: Data,
        permission: Share.Permission,
        expiresAt: Date?
    ) async throws -> Share

    /// Share a folder with a user
    /// - Parameters:
    ///   - folderId: The folder ID to share
    ///   - granteeId: User ID of the recipient
    ///   - granteePublicKeys: Recipient's public keys for key wrapping
    ///   - folderEncryptedKek: The folder's KEK encrypted for the owner (required for key wrapping)
    ///   - permission: Share permission level
    ///   - recursive: Whether to include child files and subfolders
    ///   - expiresAt: Optional expiration date
    func shareFolder(
        folderId: String,
        granteeId: String,
        granteePublicKeys: KeyManager.PublicKeys,
        folderEncryptedKek: Data,
        permission: Share.Permission,
        recursive: Bool,
        expiresAt: Date?
    ) async throws -> Share

    // MARK: - List Shares

    /// Get shares created by current user
    func getCreatedShares() async throws -> [Share]

    /// Get shares received by current user
    func getReceivedShares() async throws -> [Share]

    /// Get share details
    func getShare(shareId: String) async throws -> Share

    // MARK: - Share Actions

    /// Revoke a share (as grantor)
    func revokeShare(shareId: String) async throws

    /// Update share permission
    func updateSharePermission(
        shareId: String,
        permission: Share.Permission
    ) async throws -> Share

    /// Set or remove share expiry
    func setShareExpiry(
        shareId: String,
        expiresAt: Date?
    ) async throws -> Share

    // MARK: - Invitations

    /// Get pending share invitations
    func getInvitations() async throws -> [ShareInvitation]

    /// Accept an invitation
    func acceptInvitation(invitationId: String) async throws

    /// Decline an invitation
    func declineInvitation(invitationId: String) async throws

    // MARK: - User Search

    /// Search for users to share with
    func searchUsers(query: String) async throws -> [User]
}

File: clients/android/app/src/main/kotlin/my/ssdid/drive/domain/repository/ShareRepository.kt
package my.ssdid.drive.domain.repository

import my.ssdid.drive.domain.model.Share
import my.ssdid.drive.domain.model.SharePermission
import my.ssdid.drive.domain.model.User
import my.ssdid.drive.util.Result
import kotlinx.coroutines.flow.Flow
import java.time.Instant

/**
 * Repository interface for share operations.
 */
interface ShareRepository {

    /**
     * Get shares received by the current user.
     */
    suspend fun getReceivedShares(): Result<List<Share>>

    /**
     * Observe shares received by the current user.
     */
    fun observeReceivedShares(): Flow<List<Share>>

    /**
     * Get shares created by the current user.
     */
    suspend fun getCreatedShares(): Result<List<Share>>

    /**
     * Observe shares created by the current user.
     */
    fun observeCreatedShares(): Flow<List<Share>>

    /**
     * Get a share by ID.
     */
    suspend fun getShare(shareId: String): Result<Share>

    /**
     * Share a file with another user.
     */
    suspend fun shareFile(
        fileId: String,
        grantee: User,
        permission: SharePermission,
        expiresAt: Instant? = null
    ): Result<Share>

    /**
     * Share a file with another user by ID (for background sync).
     */
    suspend fun shareFile(
        fileId: String,
        recipientId: String,
        permission: String
    ): Result<Share>

    /**
     * Share a folder with another user.
     */
    suspend fun shareFolder(
        folderId: String,
        grantee: User,
        permission: SharePermission,
        recursive: Boolean = true,
        expiresAt: Instant? = null
    ): Result<Share>

    /**
     * Share a folder with another user by ID (for background sync).
     */
    suspend fun shareFolder(
        folderId: String,
        recipientId: String,
        permission: String
    ): Result<Share>

    /**
     * Update a share's permission level.
     */
    suspend fun updatePermission(
        shareId: String,
        permission: SharePermission
    ): Result<Share>

    /**
     * Update a share's permission level by permission string (for background sync).
     */
    suspend fun updatePermission(
        shareId: String,
        permission: String
    ): Result<Share>

    /**
     * Set or remove a share's expiry.
     */
    suspend fun setExpiry(
        shareId: String,
        expiresAt: Instant?
    ): Result<Share>

    /**
     * Revoke a share.
     */
    suspend fun revokeShare(shareId: String): Result<Unit>

    /**
     * Search for users to share with.
     */
    suspend fun searchUsers(query: String): Result<List<User>>

    /**
     * Sync shares from the server.
     */
    suspend fun syncShares(): Result<Unit>
}
